File: ui_examples/views/DatePickerEx/app/src/main/java/lee/dorian/datepicker_ex/MainActivity.java
package lee.dorian.datepicker_ex;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.CalendarView;
import android.widget.DatePicker;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private ViewHolder viewHolder;
    private ListenerHolder listenerHolder;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        viewHolder = new ViewHolder();
        listenerHolder = new ListenerHolder();
    }

    private class ViewHolder {
        public DatePicker dpExample;
        public Button btnOK;

        public ViewHolder() {
            dpExample = (DatePicker)findViewById(R.id.dpExample);
            btnOK = (Button)findViewById(R.id.btnOK);
            init();
        }

        public void init() {
            dpExample.updateDate(2018, 3 - 1, 14);
        }
    }

    private class ListenerHolder {
        public ListenerHolder() {
            viewHolder.btnOK.setOnClickListener(btnOKClickListener);
        }

        public View.OnClickListener btnOKClickListener = (clickedView) -> {
            int year = viewHolder.dpExample.getYear();
            int month = viewHolder.dpExample.getMonth();
            int day = viewHolder.dpExample.getDayOfMonth();

            Toast.makeText(
                MainActivity.this,
                String.format("%04d-%02d-%02d", year, (month + 1), day),
                Toast.LENGTH_SHORT
            ).show();
        };
    }

}

File: net_examples/image_file_download/ex01/src/main/java/lee/dorian/android/image_file_download_ex01/MainActivity.kt
package lee.dorian.android.image_file_download_ex01

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts

class MainActivity : AppCompatActivity() {

    lateinit var views: Views

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        views = Views()
    }

    inner class Views(
        val etImageURL: EditText = findViewById(R.id.etImageURL),
        val btnDownloadImage: Button = findViewById(R.id.btnDownloadimage)
    ) {
        init {
            btnDownloadImage.setOnClickListener(btnDownloadImageClickListener)
        }
    }

    val btnDownloadImageClickListener = View.OnClickListener {
        views.etImageURL.clearFocus()
        val permission = android.Manifest.permission.WRITE_EXTERNAL_STORAGE

        if (isPermissionGranted(permission)) {
            downloadImage(
                views.etImageURL.text.toString(),
                ::handlingImageDownloadResult
            )
        }
        else {
            requestPermissionLauncher.launch(permission)    // 필요 권한 요청
        }
    }

    val requestPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { isGranted ->
        if (isGranted) {
            downloadImage(
                views.etImageURL.text.toString(),
                ::handlingImageDownloadResult
            )
        }
        else {
            // 권한 부재로 이미지 다운로드 불가 알림
            Toast.makeText(this, "Can't download image due to permission denied.", Toast.LENGTH_SHORT).show()
        }
    }

    fun handlingImageDownloadResult(isSuccessful: Boolean) {
        if (isSuccessful) {
            Toast.makeText(this, "Downloaded image.", Toast.LENGTH_SHORT).show()
        }
        else {
            Toast.makeText(this, "Can't download image.", Toast.LENGTH_SHORT).show()
        }
    }
}